src/control/server/storage/config.go

```diff
@@ -259,6 +267,177 @@ func (sc *ScmConfig) Validate(class Class) error {
 	return nil
 }
 
+type stringSet map[string]struct{}
+
+func (ss stringSet) Add(s string) error {
+	if _, exists := ss[s]; exists {
+		return errors.Errorf("duplicate device %q", s)
+	}
+	ss[s] = struct{}{}
+	return nil
+}
+
+func (ss stringSet) Strings() []string {
+	var out []string
+	for s := range ss {
+		out = append(out, s)
+	}
+	sort.Strings(out)
+	return out
+}
+
+// BdevDeviceList represents a set of block device addresses.
+type BdevDeviceList struct {
+	// As this is the most common use case, we'll make the embedded type's methods
+	// available directly on the type.
+	hardware.PCIAddressSet
+
+	// As a fallback for non-PCI bdevs, maintain a map of strings.
+	stringBdevSet stringSet
+}
+
+// maybePCI does a quick check to see if a string could possibly be a PCI address.
+func maybePCI(addr string) bool {
+	comps := strings.Split(addr, ":")
+	if len(comps) != 3 {
+		return false
+	}
+	return (len(comps[0]) == 6 || len(comps[0]) == 4) && len(comps[1]) == 2 && len(comps[2]) >= 2
+}
+
+// fromStrings creates a BdevDeviceList from a list of strings.
+func (bdl *BdevDeviceList) fromStrings(addrs []string) error {
+	if bdl == nil {
+		return errors.New("nil BdevDeviceList")
+	}
+
+	if bdl.stringBdevSet == nil {
+		bdl.stringBdevSet = stringSet{}
+	}
+
+	for _, strAddr := range addrs {
+		if !maybePCI(strAddr) {
```

Comment:
here why not try to create a PCI address and if it fails then add to the string set?

Reply:
I wanted to be able to distinguish between strings that are clearly not PCI addresses (i.e. some other type of bdev) and malformed (mistyped, etc) PCI addresses in order to return the error early, here, instead of waiting until later to do the validation.